### PR TITLE
Update `sudoers` table to support newer include syntax

### PR DESCRIPTION
**Pull request by directionless**
In sudo 1.9.1, sudo changed the include directives from `#include` to
`@include` (See https://github.com/sudo-project/sudo/commit/741c6f274e65f852b7026b10a21dff519b2fd6c2)

Support parsing them.

Along the way, add more tests and update the parsing

### CODE REVIEW

File: osquery/tables/system/posix/sudoers.cpp [hunk 1]
```diff
@@ -30,6 +30,8 @@ const std::string kSudoFile = "/etc/sudoers";
 const std::string kSudoFile = "/usr/local/etc/sudoers";
 #endif
 
+const char* kSudoWhitespaceChars = "\t\v ";
```
Comment (theopolis) [inline osquery/tables/system/posix/sudoers.cpp:33]:
```suggestion
const std::string kSudoWhitespaceChars = "\t\v ";
```

can this be `const std::string` for consistency (we use `std::string` for most other consts).

File: osquery/tables/system/posix/sudoers.cpp [hunk 2]
```diff
@@ -49,50 +51,58 @@ void genSudoersFile(const std::string& filename,
...
+    // it into the leading token and the trailing token.
 
-    // Only add lines that are not comments or blank.
-    if (line.size() > 0 && line.at(0) != '#') {
-      r["source"] = filename;
+    boost::trim_if(line, boost::is_any_of(kSudoWhitespaceChars));
 
-      auto header_pos = line.find_first_of("\t\v ");
-      r["header"] = line.substr(0, header_pos);
+    if (line.size() == 0) {
```
Comment (theopolis) [inline osquery/tables/system/posix/sudoers.cpp:60]:
```suggestion
    if (line.empty()) {
```

File: osquery/tables/system/posix/sudoers.cpp [hunk 3]
```diff
@@ -49,50 +51,58 @@ void genSudoersFile(const std::string& filename,
...
-      auto space = line.find_first_of(' ');
+    // Find the next field. Instead of skipping the whitespace, we
+    // include it, and then trim it.
+    auto rule_details =
+        (header_len < line.size()) ? line.substr(header_len) : "";
+    boost::trim_if(rule_details, boost::is_any_of(kSudoWhitespaceChars));
+
+    // If an include is _missing_ the target to include, treat it like a
+    // comment.
+    if (rule_details == "" &&
```
Comment (theopolis) [inline osquery/tables/system/posix/sudoers.cpp:83]:
```suggestion
    if (rule_details.empty() &&
```

File: osquery/tables/system/posix/sudoers.cpp [hunk 4]
```diff
@@ -49,50 +51,58 @@ void genSudoersFile(const std::string& filename,
...
-      r["source"] = filename;
-      r["header"] = "#includedir";
-      r["rule_details"] = inc_dir;
-      results.push_back(std::move(r));
-
       std::vector<std::string> inc_files;
-      if (!listFilesInDirectory(inc_dir, inc_files).ok()) {
-        TLOG << "couldn't list includedir: " << inc_dir;
+      if (!listFilesInDirectory(rule_details, inc_files).ok()) {
+        TLOG << "couldn't list includedir: " << rule_details;
```
Comment (theopolis) [inline osquery/tables/system/posix/sudoers.cpp:105]:
```suggestion
        TLOG << "Could not list includedir: " << rule_details;
```

File: osquery/tables/system/posix/sudoers.cpp [hunk 5]
```diff
@@ -49,50 +51,58 @@ void genSudoersFile(const std::string& filename,
...
+    // Find the next field. Instead of skipping the whitespace, we
+    // include it, and then trim it.
+    auto rule_details =
+        (header_len < line.size()) ? line.substr(header_len) : "";
+    boost::trim_if(rule_details, boost::is_any_of(kSudoWhitespaceChars));
+
+    // If an include is _missing_ the target to include, treat it like a
+    // comment.
+    if (rule_details == "" &&
+        (header == "#include" || header == "@include" ||
```
Comment (theopolis) [inline osquery/tables/system/posix/sudoers.cpp:84]: Nitpick but if you are checking `header` here and then below, you could optimize

```
auto is_include = (header == "#include" || header == "@include");
auto is_includedir = (header == "#includedir" || header == "@includedir");
```

Then this becomes:

```
if (rule_details.empty() && (is_include || is_include_dir)) {
```

And below is either `if (is_include)` or `if (is_includedir)`.

**Review by theopolis — CHANGES_REQUESTED** [review]
A few nitpicks, then good to go

**Comment by directionless** [comment]
Thank for the feedback! I think I got all of it. I considered making `path` a context settable variable. But that seems better left to a future PR

**Review by theopolis — APPROVED** [review]

**Pull request merged**